[Docs] Add "Session Windows & On-Chain Parameters" section to claim/proof/lifecycle docs (#632)
pokt-network/poktroll: docusaurus/docs/protocol/primitives/claim_and_proof_lifecycle.md

>   - [Session Duration](#session-duration)

Author: I'm noticing that the adding a table of contents to the contents seems redundant as docusaurus builds one in the sidebar on the right.

![image](https://github.com/pokt-network/poktroll/assets/600733/036d8128-da8d-4844-acdd-eaedc400583a)

Reviewer: I still think it's useful when rendering locally.

    end
```

## Session Windows & On-Chain Parameters

_TODO(@bryanchriswhite): Add message distribution offsets/windows to this picture._

```mermaid

> gantt
>     title Session Relay / Claim / Proof Windows
>     dateFormat ss
>     axisFormat %S
>     tickInterval 1second
> 
>     section Relay Window
>         Session N Start: milestone, sns, 00, 0s
>         num_blocks_per_session: nbps, 00, 4s
>         Session N End: milestone, sne, after nbps, 0s
>         grace_period_end_offset_blocks: gpof, after sne, 2s
>         Grace Period End: milestone, gpe, after gpof, 0s
>     section Claim Window
>         claim_window_open_offset_blocks: cwob, after sne, 10ms
>         Session N Claim Window Open: milestone, cwo, after cwob, 0s
>         claim_window_close_offset_blocks: cwcb, after cwo, 4s
>         Session N Claim Window Close: milestone, cwc, after cwcb, 0s
>     section Proof Window
>         proof_window_open_offset_blocks: pwob, after cwc, 10ms
>         Session N Proof Window Open: milestone, pwo, after pwob, 0s
>         proof_window_close_offset_blocks: pwcb, after pwo, 4s
>         Session N PRoof Window Close: milestone, pwc, after pwcb, 0s

Author: ```suggestion
gantt
    title Session Relay / Claim / Proof Windows
    dateFormat ss
    axisFormat %S
    tickInterval 1second

    section Relay Window
        Session N Start: milestone, sns, 00, 0s
        num_blocks_per_session: nbps, 00, 4s
        Session N End: milestone, sne, after nbps, 0s
        grace_period_end_offset_blocks: gpof, after sne, 2s
        Grace Period End: milestone, gpe, after gpof, 0s
        Session N + 1 Start: milestone, sns1, after sne, 0s
        num_blocks_per_session: nbps2, after sns1, 4s
    section Claim Window
        claim_window_open_offset_blocks: cwob, after gpe, 10ms
        Session N Claim Window Open: milestone, cwo, after cwob, 0s
        claim_window_close_offset_blocks: cwcb, after cwo, 4s
        Session N Claim Window Close: milestone, cwc, after cwcb, 0s
    section Proof Window
        proof_window_open_offset_blocks: pwob, after cwc, 10ms
        Session N Proof Window Open: milestone, pwo, after pwob, 0s
        proof_window_close_offset_blocks: pwcb, after pwo, 4s
        Session N PRoof Window Close: milestone, pwc, after pwcb, 0s

```

    end
```

## Session Windows & On-Chain Parameters

_TODO(@bryanchriswhite): Add message distribution offsets/windows to this picture._

```mermaid
gantt
    title Session Relay / Claim / Proof Windows
    dateFormat ss
    axisFormat %S
    tickInterval 1second

    section Relay Window
        Session N Start: milestone, sns, 00, 0s
        num_blocks_per_session: nbps, 00, 4s
        Session N End: milestone, sne, after nbps, 0s

>         grace_period_end_offset_blocks: gpof, after sne, 2s

Author: ```suggestion
        grace_period_end_offset_blocks: gpof, after sne, 1s
```

        Grace Period End: milestone, gpe, after gpof, 0s
        Session N + 1 Start: milestone, sns1, after sne, 0s
        num_blocks_per_session: nbps2, after sns1, 4s
    section Claim Window

>         claim_window_open_offset_blocks: cwob, after sne, 2s

Author: ```suggestion
        claim_window_open_offset_blocks: cwob, after sne, 1s
```

>         claim_window_open_offset_blocks: cwob, after gpe, 10ms

Author: Nice catch! Must have lost this in a mermaid.live tab that I closed at some point. :sweat_smile:

        Session N Claim Window Open: milestone, cwo, after cwob, 0s
        claim_window_close_offset_blocks: cwcb, after cwo, 4s
        Session N Claim Window Close: milestone, cwc, after cwcb, 0s
    section Proof Window
        proof_window_open_offset_blocks: pwob, after cwc, 10ms
        Session N Proof Window Open: milestone, pwo, after pwob, 0s
        proof_window_close_offset_blocks: pwcb, after pwo, 4s
        Session N PRoof Window Close: milestone, pwc, after pwcb, 0s

```

> NB: Depicted with the default values (see below); x-axis is units are blocks.

| Parameter                          | Description                                                                                                                                               | Default |
|------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------|---------|
| `num_blocks_per_session`           | The number of blocks between the session start & end heights. Relays handled in these blocks are included in session N.                                   | 4       |

> | `grace_period_end_offset_blocks`   | The number of blocks after the session end height, at which the grace period ends. Relays handled in these blocks are also included in session N.         | 2       |

Author: ```suggestion
| `grace_period_end_offset_blocks`   | The number of blocks after the session end height, at which the grace period ends. Valid relays from both sessions N and N +1 are accepted in these blocks.         | 1       |
```

> | `claim_window_open_offset_blocks`  | The number of blocks after the session grace period height, at which the claim window opens. Valid claims for session N will be rejected in these blocks. This parameter MUST NOT be less than grace_period_end_offset_blocks. It is positively correlated with the number of relays in (i.e. size of) each session tree for each session number (less other scaling factors; e.g. relaymining). | 1       |

Author: > `The number of blocks after the session end height`

:+1: 

> `Valid relays from both sessions N and N +1 are accepted in these blocks.`

Isn't this describing `grace_period_end_offset_blocks`, not `claim_window_open_height`?

| `claim_window_close_offset_blocks` | The number of blocks after the claim window open height, at which the claim window closes. Valid claims for session N will be accepted in these blocks. It is negatively correlated with density of claim creation (and update) messages over blocks in a given session number.                                                                                                                  | 4       |

> | `proof_window_open_offset_blocks`  | The number of blocks after the claim window close height, at which the proof window opens. Valid proofs for session N will be rejected in these block. It is positively correlated with the amount of time suppliers MUST persist the complete merkle trees for unproven sessions (proof path is revealed at proof_window_open_height - 1).                                                      | 0       |

Reviewer 2: Maybe add a TOTO to make proof path to be revealed at earliest supplier commit height - 1

Author: I thought the idea was to reveal it as late as possible to reduce the duration between when it can be known and when it must be used?

Author: If it is revealed within the claim window (i.e. before claim window close), that's even worse as the proof path would be known before the claim (commitment) is on-chain. Am I missing something?

Reviewer 2: Added the comment w.r.t. this
> (proof path is revealed at proof_window_open_height - 1).

Author: Ah, I see it now. :dizzy_face: 

```suggestion
| `proof_window_open_offset_blocks`  | The number of blocks after the claim window close height, at which the proof window opens. Valid proofs for session N will be rejected in these block. It is positively correlated with the amount of time suppliers MUST persist the complete merkle trees for unproven sessions (proof path is revealed at earliest_supplier_proof_commit_height - 1).                                                      | 0       |
```
